icu4j/main/core/src/main/java/com/ibm/icu/message2/DateTimeFormatterFactory.java

```diff
@@ -1,31 +1,50 @@
 // © 2022 and later: Unicode, Inc. and others.
-// License & terms of use: http://www.unicode.org/copyright.html
+// License & terms of use: https://www.unicode.org/copyright.html
 
 package com.ibm.icu.message2;
 
+import com.ibm.icu.text.DateFormat;
+import java.util.Calendar;
 import java.util.Locale;
 import java.util.Map;
-import java.util.Objects;
-
-import com.ibm.icu.impl.locale.AsciiUtil;
-import com.ibm.icu.text.DateFormat;
-import com.ibm.icu.text.SimpleDateFormat;
+import java.util.TimeZone;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 /**
  * Creates a {@link Formatter} doing formatting of date / time, similar to
  * <code>{exp, date}</code> and <code>{exp, time}</code> in {@link com.ibm.icu.text.MessageFormat}.
  */
 class DateTimeFormatterFactory implements FormatterFactory {
+    private final String kind;
+
+    // "datetime", "date", "time"
+    DateTimeFormatterFactory(String kind) {
+        switch (kind) {
```

Comment:
I considered it, but it is an implementation detail, not public.
So it adds an extra type without saving much.

And `toUpperCase` is not OK.

The spec is case sensitive. `...{foo :DateTimE...}...` is a valid function name, but it is not the standard registry one.

If any of this would be public APIs I would probably do that (and use the enum to register the function in the registry, etc.)
And define the enum to do something like DATETIME("datetime"), etc.